src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Win32.cs

```diff
@@ -736,6 +739,106 @@ private static bool TryCompareTimeZoneInformationToRegistry(in TIME_ZONE_INFORMA
+        /// Searches the installed OS languages for either an exact matching culture, or one that has the same parent.
+        /// If not found, uses the preferred default OS UI language, to align with prior behavior.
+        /// </summary>
+        private static string GetCachedFileMuiPath(string filePath, CultureInfo cultureInfo)
+        {
+            string? result;
+            string cacheKey = $"{cultureInfo.Name};{filePath}";
+
+            lock (s_FileMuiPathCache)
+            {
+                if (s_FileMuiPathCache.TryGetValue(cacheKey, out result))
+                {
+                    return result;
+                }
+            }
+
+            result = GetFileMuiPath(filePath, cultureInfo);
+
+            lock (s_FileMuiPathCache)
+            {
+                s_FileMuiPathCache.TryAdd(cacheKey, result);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Helper function for getting the MUI path for a given resource and culture.
+        /// Searches the installed OS languages for either an exact matching culture, or one that has the same parent.
+        /// If not found, uses the preferred default OS UI language, to align with prior behavior.
+        /// </summary>
+        private static unsafe string GetFileMuiPath(string filePath, CultureInfo cultureInfo)
+        {
+            char* fileMuiPath = stackalloc char[Interop.Kernel32.MAX_PATH + 1];
+            char* language = stackalloc char[Interop.Kernel32.LOCALE_NAME_MAX_LENGTH + 1];
+            uint fileMuiPathLength = Interop.Kernel32.MAX_PATH;
+            uint languageLength = Interop.Kernel32.LOCALE_NAME_MAX_LENGTH;
+            ulong enumerator = 0;
+
+            while (true)
+            {
+                // Search all installed languages.  The enumerator is re-used between loop iterations.
+                bool succeeded = Interop.Kernel32.GetFileMUIPath(
+                    Interop.Kernel32.MUI_USE_INSTALLED_LANGUAGES,
+                    filePath, language, ref languageLength,
+                    fileMuiPath, ref fileMuiPathLength, ref enumerator);
+
+                if (!succeeded)
+                {
+                    // Recurse to search using the parent of the desired culture.
+                    if (cultureInfo.Parent.Name != string.Empty)
+                    {
+                        return GetFileMuiPath(filePath, cultureInfo.Parent);
+                    }
+
+                    // Final fallback, using the preferred installed UI language.
+                    enumerator = 0;
+                    succeeded = Interop.Kernel32.GetFileMUIPath(
+                        Interop.Kernel32.MUI_USER_PREFERRED_UI_LANGUAGES,
+                        filePath, language, ref languageLength,
+                        fileMuiPath, ref fileMuiPathLength, ref enumerator);
+
+                    if (succeeded)
+                    {
+                        fileMuiPath[Interop.Kernel32.MAX_PATH] = '\0';
```

Comment:
This line exists in multiple places here and below.  It would be safer and simpler to just dedup that into a single line above just after the GetFileMUIPath call.

Reply:
Not sure if it's even needed.  The API always returns a null terminated string.  Tarek asked for it just to be safe.  Note, the two length values from the API are useless - they are the buffer lengths, not the values lengths.